### DISCUSSION TITLE
Doubt regarding matrices returned by collocation_coeff()

### DISCUSSION THREAD

**Discussion by itbellix** [casadi/casadi#3269] [Q&A] [answered]
Hi CasADi developers/users,
I am going  through the [`direct_collocation_opti.m`](https://github.com/casadi/casadi/blob/main/docs/examples/matlab/direct_collocation_opti.m) example and I see that it makes use of the `collocation_coeff()` function, as opposed to the [`direct_collocation.py`](https://github.com/casadi/casadi/blob/main/docs/examples/python/direct_collocation.py) example where the matrices of the coefficients for the continuity and collocation equations (and quadratures) are calculated explicitly. While I think this could just be for "historical" reasons, such as the fact that the `collocation_coeff()` function came after that tutorial was written, I am a bit confused by the fact that the matrices that are obtained in the two examples are different. In particular, I see that in the [last lines of the CPP implementation of `collocation_coeff()`](https://github.com/casadi/casadi/blob/b3ccf30fc2072933a5f3e797989554525a5c2dc9/casadi/core/integration_tools.cpp#L283C1-L286C55) the continuity equation coefficients are returned also for the collocation point `0`, while this does not happen for the collocation equation and for the quadrature (that are sliced). 
Is my understanding correct? What is the reason behind this? 

A simple follow-up observation would be that that, because of the different implementation, it is in my opinion quite difficult for users replace the "direct computation" of the collocation coefficients with the `collocation_coeff()` function, as this requires understanding exactly how the output matrices differ in the two cases.

**Comment by jgillis** [comment, marked as answer]
Hi @itbellix, I'm not aware of standard terminology or semantics for these coefficient matrices; basically it's arbitrary.

Looking up the answer, I did discover that the documentation seems to be missing online even though it is present in the headers ( https://github.com/casadi/casadi/blob/main/casadi/core/integration_tools.hpp#L79-L120 and through `help casadi.collocation_coeff`).

In collocation methods, you don't enforce the derivative constraint at the start of the collocation interval. That would result in an overdefined system of equations. Hence the C matrix as defined in `direct_collocation.m` has a first column that is never used.

`collocation_coeff` was really designed to obtain the least clunky-looking implementation of `direct_collocation_opti` without regard for direct computation.

**Comment by itbellix** [comment]
Hi @jgillis , thank you!
My doubt was indeed stemming from the fact that in one of the two examples derivatives at the beginning of the collocation interval were computed (but as you say, not used) while in the second this was not the case.

Thank you for your answer and time.
